### ISSUE TITLE
Support restricting core/std to a particular version (for msrv purposes)

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/rust#108406]
Since all library types/functions are tagged with the particulars of the rust release that saw their stabilization, I propose that it should be possible to pass a value to `rustc` to have it "pretend" to be an older version by refusing to compile with any features that wouldn't have been available as of a particular msrv version.

Currently projects use a `rust-toolchain.toml` or the `RUSTUP_TOOLCHAIN` environment variable in conjunction with a rustup-provided `rustc` shim to make sure that a project is developed/tested with a particular version (or branch) of the toolchain but this has a number of drawbacks, not the least of which is heavily penalizing OSS developers that contribute to multiple projects each using `rust-toolchain.toml` to enforce some MSRV by requiring them to install each and every rust msrv version.

Obviously this is not intended to be a stand-in for actually compiling and testing code w/ older rust releases to ensure that other issues, deprecations, subtle changes in implementation, etc. all don't affect the code but it would be great if testing (or enforcing) compatibility with a particular version of rust were as simple as using `rustc --msrv xx.yy ..` instead of requiring installation of a separate toolchain altogether.

There's an argument to be made for also allowing the nightly toolchain to pretend to be stable (by disallowing feature gates, etc.) but I imagine the distance between "actual previous-release behavior" and "mock msrv behavior" would necessarily be larger if trying to pretend that rustc-nightly was actually rustc-xx.yy-stable, so I'd be content with just the former :)

**Comment by workingjubilee** [comment]
That won't cover language features which are stabilized that don't affect the surface implementations of types.

**Comment by mqudsi** [comment]
Yes, that is correct. It would not be a complete stand-in for actually testing with an older version.

**Comment by workingjubilee** [comment]
I think There's a lot more to fix in Rustup, like enforcing version normalization so that 3 projects, each with a rust-toolchain.toml, but variantly spelling the version strings:
- "1.67"
- "1.67.1"
- "stable"
doesn't require 3 toolchains.
